### PR TITLE
use db query results to render timelines

### PR DESCRIPTION
**Pull request by krsnik93**
Incorporates query results into rendering timelines

### CODE REVIEW

File: pepys_timeline/static/js/src/pepys.js [hunk 1]
```diff
@@ -57,17 +61,70 @@ function onDataReceived() {
...
 }
 
+function fetchConfig() {
+    fetch('/config')
+        .then(response => response.json())
+        .then(response => {
+            const { frequency_secs } = response;
+            fetchSerials();
+            fetchSerialsMeta();
+            setInterval(fetchSerials, frequency_secs * 100);
```
Comment (IanMayo) [inline pepys_timeline/static/js/src/pepys.js]: this needs to be 1000 to convert to millis @krsnik93

File: pepys_timeline/static/js/src/pepys.js [hunk 2]
```diff
@@ -179,49 +300,54 @@ function renderCharts(serials) {
...
-
-  const serial_participants = [{
-    "serial_id": "faa18ef7-6823-33dc-0f16-de6e4b2c02f3",
-    "platform_id": "50d64387-9f91-4c6f-8933-f4e7b1a7d8ab",
-    "start": "2020-11-15 00:00:00",
-    "end": "2020-11-15 23:59:59",
-    "gap_seconds": 30
-  }];
-  const range_types = ["G", "C"];
+            // take deep copy of data. For some reason using a dataset
```
Comment (IanMayo) [inline pepys_timeline/static/js/src/pepys.js]: I'm pretty sure we don't need to do this deep copy any more @krsnik93 , we're not re-using mock datasets any more.

**Comment by IanMayo** [comment]
Nice work @krsnik93  - great JS array processing skills :-)

**Comment by IanMayo** [comment]
Minor missing feature.  The participants are split into two forces, red & blue.  Here is the _current_ database data:
image

So, when listing the participants we need to show the Blue forces, then the Red forces (with appropriately shaded icon).

**Comment by IanMayo** [comment]
This is still a draft PR - so I know it's incomplete @krsnik93 .  Just a little more feedback.  The data doesn't seem to match the data in Tracstor.   Are we still using dummy data, or are you connecting to another database?

(just wondering...)

**Comment by krsnik93** [comment]
@IanMayo I will update tonight regarding red/blue participants, definitely missing this right now.

**Comment by krsnik93** [comment]
@arunlal-v could we include force type (including name and color) in the `metadata` sql result?

**Comment by rnllv** [comment]
> @arunlal-v could we include force type (including name and color) in the `metadata` sql result?

Sure, I'll work on that.

**Comment by rnllv** [comment]
Changes are now available in trac-store1 db.

**Comment by IanMayo** [comment]
Hello @krsnik93  - sorry to interrupt :-)

I'm trying to gain an understanding of how the process works.  For the`dashboard_stats` query, are you passing all of the `dashboard_metadata` results in the query?
I was just expecting discrete fields, for things like: `serial_reference`, `platform_id`, plus maybe `start_time` and `end_time`.

**Comment by krsnik93** [comment]
@IanMayo I filter away serials from the `dashboard_metadata` result and pass the serial participant dictionaries/jsons unchanged into the `dashboard_stats`. If you are concerned about performance or security, I can filter down fields.

**Comment by IanMayo** [comment]
Thanks for that @krsnik93 .  It's the diagnostics that are the challenge ;-). If we have any issues, they will be easier to diagnose if we pass the minimum number of fields.  Inspecting the calls in the Network tab of Chrome Developer tools is particularly challenging.

**Comment by IanMayo** [comment]
I wonder if we have a cache-ing issue.

I saw that the two serials both had the same participants. That's no surprise, since I created the second by making a duplicate of the first.  But, having the same participants in both lists makes it harder to check things are working.
Screen Shot 2021-04-28 at 16 20 16


So, I deleted several participants for one serial.  But 10 minutes later the were still showing as 6 participants each.  I did <ctrl-r> hard refresh, and when the ui opened, it showed different length lists.

Screen Shot 2021-04-28 at 16 20 32

**Comment by krsnik93** [comment]
@IanMayo I have added a flask extension for cache busting. Let's see if we still get this behaviour and whether performance drops.

**Comment by krsnik93** [comment]
Definitely seems slower on initial load, but this might not matter that much for our use case.

**Comment by IanMayo** [comment]
> slower on initial load
Sure, that's not a great issue.

On the current `deployment` from this branch I'm getting this JS error on the one-minute update:
image

**Comment by krsnik93** [comment]
should be fixed now... I will look into the `NaN` value for one of the serials

**Comment by IanMayo** [comment]
@krsnik93 - I just changed the participation time for `SALT` in `J05250`.  On the next update it correctly moved over.

I also changed the name of `J05051`, but it wasn't reflected in the data. I guess you're not looking out for something like that changing. The `serial_id` GUID won't change, but the `reference` may - particularly if it has been mis-typed.

I tried changing the name of a participant, and the timeline updated correctly.

**Comment by IanMayo** [comment]
Hello @krsnik93 
we now have measurements for two platforms in the data. I _think_ the coverage/gaps calculation is mistaken (#18), but it does mean you have some _real_ data to play with, and consider the rendering.

**Comment by IanMayo** [comment]
Note: I just (temporarily) included another Serial in the results, but the phase-2 query then failed because it was too long:
image

I guess this will be fixed once the query is trimmed down to the essential parameters.

**Comment by krsnik93** [comment]
> Hello @krsnik93 - sorry to interrupt :-)
> 
> I'm trying to gain an understanding of how the process works. For the`dashboard_stats` query, are you passing all of the `dashboard_metadata` results in the query?
> I was just expecting discrete fields, for things like: `serial_reference`, `platform_id`, plus maybe `start_time` and `end_time`.

addressed here: https://github.com/debrief/pepys-timeline/pull/17/commits/e401f809e2d1b1d892d3589304ce1a2d8c071124 to send only what's required here: https://github.com/debrief/pepys-timeline/blob/b17ef1b6f4c8547e7a4df66af6b2c78181c218d9/sql/dashboard_stats.sql#L39

**Comment by krsnik93** [comment]
@IanMayo using force type colors now:
image

Still need to sort by force. I would suggest a different set of colors in the database.

**Comment by krsnik93** [comment]
@IanMayo what sort order would you suggest? First by force type (blue, then red), then alphabetically by participant name?

**Comment by IanMayo** [comment]
> First by force type (blue, then red), then alphabetically by participant name?

Exactly correct :-)

**Comment by krsnik93** [comment]
image

**Comment by IanMayo** [comment]
@krsnik93 - I've been playing with your branch ;-)

As discussed in [Slack](https://pepys-import.slack.com/archives/CU09ELTNK/p1619637660017000) we need to apply a 5x scaling factor to the interval time.

I've amended that (possibly a bit of a botch-job, feel free to improve it), and now we see valid gap data:
https://pepys-import.slack.com/archives/CU09ELTNK/p1619637660017000

This is more representative of _real_ data, and the task of calculating coverage should be easier now.

image

**Comment by rnllv** [comment]
> Note: I just (temporarily) included another Serial in the results, but the phase-2 query then failed because it was too long:
> image
> 
> I guess this will be fixed once the query is trimmed down to the essential parameters.

@krsnik93, just curious as to why we are preferring HTTP GET method here. The json structure can change in the future, and would make the url longer. Wouldn't it be performant to use HTTP POST here since json is natively supported  by HTTP as a content-type?

**Comment by krsnik93** [comment]
Usually POST is used for creating resources and GET for searching and getting them. However, it looks like the length is causing issues anyways so we can switch to POST.

**Comment by krsnik93** [comment]
closing this as we now want to merge into `main`, `UI_Prototyping`.

**Pull request closed without merging**